crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1804,70 +1807,145 @@ impl<'db> ClassType<'db> {
+        let has_type_params = class_stmt.type_params.is_some();
+
+        class_stmt
             .bases()
             .iter()
-            .map(move |base_expr: &ast::Expr| {
-                if class_stmt_node.type_params.is_some() {
-                    // when we have a specialized scope, we'll look up the inference
-                    // within that scope
-                    let model: SemanticModel<'db> = SemanticModel::new(db, definition.file(db));
-                    base_expr.ty(&model)
-                } else {
-                    // Otherwise, we can do the lookup based on the definition scope
-                    definition_expression_ty(db, definition, base_expr)
-                }
-            })
+            .map(move |base_node| infer_class_base_type(db, base_node, definition, has_type_params))
+    }
+
+    /// Return the original [`ast::StmtClassDef`] node associated with this class
+    fn node(self, db: &'db dyn Db) -> &'db ast::StmtClassDef {
+        match self.definition(db).kind(db) {
+            DefinitionKind::Class(class_stmt_node) => class_stmt_node,
+            _ => unreachable!("Class type definition should always have DefinitionKind::Class"),
+        }
+    }
+
+    /// Attempt to resolve the [method resolution order] ("MRO") for this class.
+    /// If the MRO is unresolvable, return an error indicating why the class's MRO
+    /// cannot be accurately determined.
+    ///
+    /// The MRO is the tuple of classes that can be retrieved as the `__mro__`
+    /// attribute on a class at runtime.
+    ///
+    /// [method resolution order]: https://docs.python.org/3/glossary.html#term-method-resolution-order
+    #[salsa::tracked(return_ref)]
+    fn try_mro(self, db: &'db dyn Db) -> Result<Mro<'db>, MroError<'db>> {
+        Mro::of_class(db, self)
+    }
+
+    /// Iterate over the [method resolution order] ("MRO") of the class.
+    ///
+    /// If the MRO could not be accurately resolved, this method falls back to
+    /// an MRO that has the class directly inheriting from `Unknown`. Use
+    /// [`ClassType::try_mro`] if you need to distinguish between the success and failure
+    /// cases rather than simply iterating over the inferred resolution order for the class.
+    ///
+    /// [method resolution order]: https://docs.python.org/3/glossary.html#term-method-resolution-order
+    fn mro(self, db: &'db dyn Db) -> impl Iterator<Item = ClassBase<'db>> {
+        // We avoid materialising the *full* MRO unless it is actually necessary:
+        //
+        // - Materialising the full MRO is expensive
+        // - We need to do it for every class in the code that we're checking, as we need to make sure
+        //   that there are no class definitions in the code we're checking that would cause an
+        //   exception to be raised at runtime. But the same does *not* apply for every class
+        //   in third-party and stdlib dependencies: we never emit diagnostics about non-first-party code.
```

Comment:
While it's true that we don't emit diagnostics for those, we still call `check_class_definitions` for every `ClassDefinitoion` which results in eagerly pulling the class's MRO. 

That's why I don't think the optimization is really about avoiding computing the MRO. It's only about avoiding making the salsa query call.